### DISCUSSION TITLE
Unable to map boolean field

### DISCUSSION THREAD

**Discussion by bardale** [finos/TimeBase-CE#283] [Q&A]
Greeting @alex-karpovich 

I am able to map every field from stream to Message class except boolean field. 
Please find below sample stream and code for reference. 

**Stream**: 
```shell
DURABLE STREAM "boolean_test" (
    CLASS "com.models.BooleanTest" 'Boolean Test' (
        "venue" 'Venue' VARCHAR,
        "on_venue" 'On venue' BOOLEAN
    );
)
OPTIONS (FIXEDTYPE; PERIODICITY = 'IRREGULAR'; HIGHAVAILABILITY = FALSE; UNIQUE = FALSE; STORAGEVERSION = '5')
COMMENT 'boolean_test'

``` 
**Data**:
<html xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:w="urn:schemas-microsoft-com:office:word"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns:m="http://schemas.microsoft.com/office/2004/12/omml"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Word.Document>
<meta name=Generator content="Microsoft Word 15">
<meta name=Originator content="Microsoft Word 15">
<link rel=File-List
href="file:///C:/Users/barddee/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">

<link rel=themeData
href="file:///C:/Users/barddee/AppData/Local/Temp/msohtmlclip1/01/clip_themedata.thmx">
<link rel=colorSchemeMapping
href="file:///C:/Users/barddee/AppData/Local/Temp/msohtmlclip1/01/clip_colorschememapping.xml">

<style>

</style>

</head>

<body lang=EN-US style='tab-interval:.5in;word-wrap:break-word'>

symbol | timestamp | BooleanTest.venue | BooleanTest.on_venue |  
-- | -- | -- | -- | --
Symbol1 | 2022-03-05T12:16:40.000Z | Venue1 | TRUE |  
Symbol1 | 2022-03-05T12:33:20.000Z | Venue2 | FALSE |  
Symbol2 | 2022-03-05T15:36:40.000Z | Venue3 |   |  

</body>

</html>

**Corresponding Message Class:**

`@SchemaElement(name = "boolean_test")
public class BooleanTestMessage2 extends InstrumentMessage {

    @SchemaElement(name = "venue")
    public String venue;

@SchemaElement(name = "on_venue")
public boolean onVenue;

}
`
**Sample code to read data:**

`
public class TestBoolean {
    public static void main(String[] args) {

        DXTickDB db = TickDBFactory.createFromUrl ("dxtick://XXXXXXX:8011");
        db.open (true);       

        SelectionOptions options = new SelectionOptions();

        options.typeLoader = new SimpleTypeLoader(null, BooleanTestMessage.class) {
            @Override
            public Class<?> load(ClassDescriptor cd) throws ClassNotFoundException {
                if (cd.getName().startsWith("QUERY"))
                    return BooleanTestMessage.class;
                return super.load(cd);
            }
        };
        options.raw = false;

        String queryF = "select  venue, on_venue from boolean_test  where timestamp BETWEEN '2022-03-05 00:00:00.000'd AND '2022-03-05 12:33:59.000'd";

        try (InstrumentMessageSource cursor = db.executeQuery(queryF, options)){
            while (cursor.next()) {
                BooleanTestMessage testMessage = (BooleanTestMessage) cursor.getMessage();
                System.out.println(testMessage);
            }
        } finally {
            db.close();
        }
}

}
`

I am able to map venue, symbol and timestamp from boolean_test stream to BooleanTestMessage except on_venue field. 
Cursor.getMessage is not showing values of "on_venue" field from Stream. 

Can you help to resolve this issue? 

Regards,
Deepak

**Comment by bardale** [comment]
Hello @alex-karpovich 
Can you please help me to resolve issue or provide some code for reference?

**Comment by alex-karpovich** [comment]
Greeting, @bardale 

here is correct sample:
```
 @SchemaElement(name = "com.models.BooleanTest")
    public static class BooleanTestMessage extends InstrumentMessage {

        @SchemaElement(name = "venue")
        public String venue;

        @SchemaElement(name = "on_venue")
        public byte onVenue;

        @Override
        public String toString() {
            StringBuilder str = new StringBuilder();
            str.append("{ \"$type\": \"BooleanTestMessage\"");

            if (hasSymbol())
                str.append(", \"symbol\": \"").append(symbol).append("\"");

            if (hasTimeStampMs())
                str.append(", \"timestamp\": \"").append(getTimeString()).append("\"");

            str.append(", \"venue\": \"").append(venue).append("\"");
            str.append(", \"onVenue\": \"").append((onVenue == 1 ? true : (onVenue == 0 ? false : "<null>"))).append("\"");

            str.append("}");
            return str.toString();
        }
    }
```

when you do query "select ..." all fields becomes nullable in resulting schema, that's why you need to have field declared as `byte` to contain 3 values, including nulls = BooleanDataType.NULL

https://github.com/finos/TimeBase-CE/blob/0077825714334fa8012dd34ea564a290a4574af1/java/timebase/pub/src/main/java/com/epam/deltix/qsrv/hf/pub/md/BooleanDataType.java#L30
